### DISCUSSION TITLE
Update parts of the UI when authenticated with Supabase and NextJS 14

### DISCUSSION THREAD

**Discussion by jvlobo** [supabase/supabase#21573] [Questions]
I'm using NextJS 14 (App Router) and Supabase Auth (@supabase/ssr)

The authentication is working all perfectly fine, but I have parts of my UI that I would like to update when the user is authenticated. For example, I have a header.tsx component with a top bar that should show "Log In" button when the user is not authenticated and a "Log out" one when the user is. However, the state is not changing automatically when the user authenticates. If I refresh the window it works well.

I've seen there is a listener supabase.auth.onAuthStateChange that you can use to listen to auth state changes but it doesn't seem to work for me.

This is my header.tsx component:

```
'use client'

import { useEffect, useState } from 'react';
import { useRouter } from 'next/navigation';
import { createClient } from '@/utils/supabase/client'

export default function Header() {
  const router = useRouter()
  const supabase = createClient()
  const [userIsLogged, setuserIsLogged] = useState<boolean>(false)

  useEffect(() => {
    supabase.auth.onAuthStateChange((event, session) => {
      console.log('onAuthStateChange', event, session)
    })

    const loadSession = async () => {
      const { data: { session }, error } = await supabase.auth.getSession()
      if (error) {
        console.error(error)
      }

      if (session) {
        setuserIsLogged(true)
      }
    }

    loadSession()
  }, [])


  return (
    <div>
      {userIsLogged ? (
        <form action="/auth/signout" method="post">
          <button type="submit">Sign Out</button>
        </form>
      ) : (
        <button onClick={() => router.push('/auth')}>Log in</button>
      )}
    </div>
  );
}
```

I have also tried returning the unsubscribe of the auth listener as I've seen recommended somewhere but didn't work

For creating my login screen I followed the Supabase docs -> https://supabase.com/docs/guides/auth/server-side/nextjs
So I have everything like it is explained there, my form, my actions, the Supabase client etc...

Any ideas on what I'm doing wrong or what I can try to get it to work?

Thanks!

**Comment by thedevdavid** [comment]
I copied the [user context provider](https://github.com/supabase/auth-helpers/blob/main/packages/react/src/components/SessionContext.tsx) from auth-helpers/react package and made some app-specific changes. 

You should be able to use it without any changes. Wrap your header in `SessionContextProvider` and then you can use the `useSession` or `useUser` hooks in `header`.

**Comment by genthegreat** [comment]
I have a similar issue.

**Comment by genthegreat** [comment]
I copied a AuthContext from here https://gist.github.com/chadmuro/b1c398d68da90eeac70bf1ca01d4a822

it seems to work for me
`const { user } = useAuth();`

then use the variable for conditional rendering

**Comment by Abraham599** [comment]
@jvlobo could you please share your workaround of this issue ?

I am stuck with the same problem.
